### PR TITLE
Sync newly returned report objects from commands

### PR DESCRIPTION
**Pull request by jasperhuangg**
### Details
We were re-fetching the chat and IOU report objects once the payIOU, payWithWallet, and rejectTransaction commands completed, which was completely unnecessary. Now, since we're returning the updated IOU and Chat report objects from https://github.com/Expensify/Web-Expensify/pull/31190, we can directly update Onyx with them.

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/165863

### Tests
Make sure you have https://github.com/Expensify/Web-Expensify/pull/31318 either pulled or deployed.
1. Create accounts A and B (or use any two testing accounts with no IOU report history between one another)
2. Use A to request money from B.
3. In account B, use "I'll settle up elsewhere" to pay the IOU report.
4. Verify that the IOU report is settled, and the UI is updated to reflect it:
Screen Shot 2021-06-15 at 6 13 22 PM
4. Use A to request money from B.
5. In account B, decline the request.
6. Verify that the transaction was rejected correctly, and that the UI is updated to reflect it:
Screen Shot 2021-06-15 at 6 15 17 PM

### QA Steps
Same as the above tests

### Tested On
- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots

https://user-images.githubusercontent.com/31285285/122041383-7a7e6d00-ce0b-11eb-98da-df9f4dcfcf47.mp4

https://user-images.githubusercontent.com/31285285/122041403-7f432100-ce0b-11eb-80b9-2ecc085305a6.mp4

https://user-images.githubusercontent.com/31285285/122041408-80744e00-ce0b-11eb-8448-b19a0db1aa59.mp4

https://user-images.githubusercontent.com/31285285/122041422-823e1180-ce0b-11eb-8956-79aee92dde3d.mp4

### CODE REVIEW

**Comment by Julesssss** [comment]
I'm happy to review this as an additional reviewer, but we should always assign Pullerbear :)

**Comment by mountiny** [comment]
I was testing it with the Web-Expensify brach and I was getting this [stacktrace error](https://github.com/Expensify/Web-Expensify/pull/31190#issuecomment-863896464) and the create report button was just spinning for very long time. Not sure what to do about it, seems like Jules pointed at some problem with returning information about the Report, which could be the reason why it keeps spinning.

**Comment by Julesssss** [comment]
> Not sure what to do about it, seems like Jules pointed at some problem with returning information about the Report, which could be the reason why it keeps spinning.

In my case, I didn't see the spinner. The request returned quickly, but it was missing the extra data 😕

**Comment by mountiny** [comment]
To be fair, there is something wrong with my dev environment, it is a second day it does not work as it should. I will escalate this to #engineering-chat

File: src/libs/actions/IOU.js
Comment (mountiny) [inline src/libs/actions/IOU.js:148]: NAB for consistency use capitalized IOU in this line of comment as well
```suggestion
            // Save the updated chat and IOU reports sent back in the response
```

File: src/libs/actions/IOU.js
Comment (mountiny) [inline src/libs/actions/IOU.js:209]: NAB same as above. Let's keep the abbreviation uppercase.
```suggestion
            // Save the updated chat and IOU reports sent back in the response
```

File: src/libs/actions/Report.js
Comment (mountiny) [inline src/libs/actions/Report.js:1284]:
```suggestion
    chatReportData[chatReportKey].hasOutstandingIOU = iouReportStuff.stateNum
        === (CONST.REPORT.STATE_NUM.PROCESSING && iouReportStuff.total !== 0);
```

Let's put it into parantheses to make it clear these two conditional belong together.

File: src/libs/actions/Report.js
Comment (mountiny) [inline src/libs/actions/Report.js:1272]:
```suggestion
 * and they've both been updated in the back-end.
 
 * @param {Object} chatReportStuff
```
NAB adding one blank line could be beneficial in terms of readability and consistency with the other functions.

**Review by mountiny — DISMISSED** [review]
Tested the branch with the relevant Web-E branch and it works well. I will approve since my comments are NAB, but consider implementing them and re-requesting the review. Good job with both of those PRs Jasper!

**Comment by jasperhuangg** [comment]
Thanks for the suggestions @mountiny, requested another review!

**Review by mountiny — DISMISSED** [review]
Great @jasperhuangg!

**Comment by Julesssss** [comment]
THis is looking good, but I believe we're [still held on this issue](https://github.com/Expensify/Web-Expensify/pull/31190#pullrequestreview-685435854) in the Web-E PR?

**Review by Julesssss — CHANGES_REQUESTED** [review]
The API tests well, but I think something is off here. When rejecting a transaction the IOU Preview remains and shows a total. But the total should be 0, and the Preview should be hidden. 

This only occurred for me when I re-opened the Modal, as this retrieved the IOU data again.

**Comment by mountiny** [comment]
I have taken my approval down and I will test it again once the problem mentioned by Jules will be addressed.

**Comment by jasperhuangg** [comment]
> The API tests well, but I think something is off here. When rejecting a transaction the IOU Preview remains and shows a total. But the total should be 0, and the Preview should be hidden.
> 
> This only occurred for me when I re-opened the Modal, as this retrieved the IOU data again.

Ah yup, this was the bug I was asking you about before. I wasn't sure what state the report was supposed to be in after it was rejected, and I guess this isn't correct.

**Comment by Julesssss** [comment]
Hey @jasperhuangg, sorry to be a pain but I think I was wrong previously. I tested this branch again today against Web-E master and everything works as expected...

> The API tests well, but I think something is off here. When rejecting a transaction the IOU Preview remains and shows a total. But the total should be 0, and the Preview should be hidden.

I can no longer reproduce this. Are you still able to reproduce this problem? If not, then I think we're all good here.

**Comment by jasperhuangg** [comment]
@Julesssss It's still happening on my end with Web-E master. Here's a screen recording of how it looks. It's a bit hard to see, but as soon as the API.RejectTransaction returns with the report object, we update the UI incorrectly. If I reload the IOU report, then everything works as expected.

https://user-images.githubusercontent.com/31285285/123364724-89051b00-d5a7-11eb-8882-5c9f9be95d4e.mp4

The initial value returned from API.RejectTransaction returns a total !== 0. This is what's causing the UI to update incorrectly. Once I reload the IOU report (as in the video), then an IOU report with total === 0 is returned

For whatever reason, [this Get call](https://github.com/Expensify/Web-Expensify/blob/0b7ceb20ca91535bc771525abd3d5274a907bf26/lib/ReportAPI.php#L1511-L1520) is getting an IOU report with a total that hasn't been updated.

**Comment by Julesssss** [comment]
This is unblocked now, testing.

File: src/libs/actions/IOU.js
Comment (Julesssss) [inline src/libs/actions/IOU.js:146]: I think `stuff` complicates the variable names. `chatReport` & `iouReport` should be fine.

File: src/libs/actions/IOU.js
Comment (Julesssss) [inline src/libs/actions/IOU.js:213]: Same here, lets just call them what they are

File: src/libs/actions/Report.js
Comment (Julesssss) [inline src/libs/actions/Report.js:1274]: Can you add a bit of the comment from [here](https://github.com/Expensify/Expensify.cash/pull/3544/files#diff-17f969ae17c2a0f476494263758dbe0a535432f496b805345925e648d605915cR206-R210)? I think this is where the main comment explanation should be, rather than duplicated in the two places it is called.

File: src/libs/actions/IOU.js
Comment (Julesssss) [reply src/libs/actions/IOU.js:146]: We're not using `chatReportStuff` anywhere else

File: src/libs/actions/Report.js
Comment (Julesssss) [inline src/libs/actions/Report.js:1277]: `chatReport`
`iouReport`

File: src/libs/actions/Report.js
Comment (Julesssss) [inline src/libs/actions/Report.js:1281]: `simplifiedIouReport` & `simplifiedReport`?

File: src/libs/actions/Report.js
Comment (Julesssss) [inline src/libs/actions/Report.js:1279]: What happens if one of the reports is empty, due to a backend issue? Could we check for null reports and return early as a basic protection?

Comment (Julesssss) [reply src/libs/actions/Report.js:1279]: I worry that in this case we'll be putting bad data into Onyx.

File: src/libs/actions/Report.js
Comment (Julesssss) [inline src/libs/actions/Report.js:1286]: I think this key should not be set if the linkedIOU report is closed (after PayIOU for example).

**Review by Julesssss — CHANGES_REQUESTED** [review]
It tests well, but I made a few suggestions.

**Comment by mountiny** [comment]
Tested it too, works as expected 👏

File: src/libs/actions/IOU.js
Comment (mountiny) [inline src/libs/actions/IOU.js:141]: NAB grammar
```suggestion
            // we also need to set the iouReportID and hasOutstandingIOU fields of the chatReport in Onyx manually.
```

File: src/libs/actions/IOU.js
Comment (mountiny) [inline src/libs/actions/IOU.js:208]: NAB grammar
```suggestion
            // Save the updated chat and IOU reports sent back in the response.
            // NOTE: Since the API doesn't handle syncing chat reports with IOU reports,
            // we also need to set the iouReportID and hasOutstandingIOU fields of the chatReport in Onyx manually.
```

File: src/libs/actions/IOU.js
Comment (mountiny) [inline src/libs/actions/IOU.js:140]: NAB grammar
```suggestion
            // Save the updated chat and IOU reports sent back in the response.
            // NOTE: Since the API doesn't handle syncing chat reports with IOU reports,
```

**Review by mountiny — COMMENTED** [review]
NAB: In addition to Jules' comments, just a small grammar details to keep the comments consistent.

File: src/libs/actions/Report.js
Comment (jasperhuangg) [reply src/libs/actions/Report.js:1286]: Not too sure what you mean here by linkedIOU report. Are you referring to the iouReport that already exists in Onyx?

Comment (Julesssss) [reply src/libs/actions/Report.js]: >  Are you referring to the iouReport that already exists in Onyx?

Yep. Currently this is not a problem because we use `hasOutstandingIOU`, but really the chatReport should no longer be linked to the paid iouReport. It's not accurate and potentially confusing. 

But I just realised that this change wasn't introduced by your PR, so no need to make this change now.

File: src/libs/actions/IOU.js
Comment (Julesssss) [inline src/libs/actions/IOU.js]: Was this supposed to be removed? Am I missing something, or was this a mistake?

**Review by Julesssss — CHANGES_REQUESTED** [review]
Changes look good, but I have [one outstanding comment](https://github.com/Expensify/Expensify.cash/pull/3544/files/1b4f5e6cfc2db61d9b8ca1a6de2f811b842aef3e..e5ea089cfe8f0a5dff36bccd856989c78ea49266#r662129401)

**Review by Julesssss — APPROVED** [review]
My comment was referring to a commit that isn't included. Good to merge once `Web-Expensify/pull/31318` is deployed to prod 👍

**Review by mountiny — APPROVED** [review]
Great! Let's wait for the deploy before merging.

**Comment by jasperhuangg** [comment]
Looks like https://github.com/Expensify/Web-Expensify/pull/31318 has been deployed, so I'm gonna merge this!

**Comment by roryabraham** [comment]
Unfortunately the deploy comments are not working right now. This was deployed to staging yesterday.

**Pull request merged**
